### PR TITLE
Issue/6025 implement top level handler abc

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description

Part of https://github.com/inmanta/inmanta-core/issues/6025

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:58]: Don't forget about this TODO

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py
Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py:826]: I guess you mean `CRUDHandler` ? This would mean we need to lift some functionality from ResourceHandler to HandlerABC ?

Comment (sanderr) [reply src/inmanta/agent/handler.py:826]: No, I mean `ResourceHandler`. If it inherits from `HandlerABC`, `CRUDHandler` doesn't need to. Now we have
```python
class HandlerABC: pass
class ResourceHandler: pass
class CRUDHandler(ResourceHandler, HandlerABC): pass
```

and I'm proposing
```python
class HandlerABC: pass
class ResourceHandler(HandlerABC): pass
class CRUDHandler(ResourceHandler): pass
```

Comment (sanderr) [reply src/inmanta/agent/handler.py:826]: My first snippet is not right. Not sure if you changed it or I saw it wrong. Anyway, if `ResourceHandler(HandlerABC)`, why would we explicitly inherit from it in `CRUDHandler`?

Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py:826]: Right, 100% agree

**Review by bartv — COMMENTED** [review]
Not sure I can contribute here. I a missing missing the point on what the goal of this refactor is.

File: src/inmanta/agent/handler.py
Comment (arnaudsjs) [inline src/inmanta/agent/handler.py:463]: The docstrings on these methods have to be improved. What is the behavioral difference between `deploy()` and `execute()` for example? It's not possible to derive that information for the name itself.

Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py:463]: I agree, I was planning on hoisting these methods from `ResourceHandler` into `HandlerABC`, but now that I've worked on the actual implementation of the `DiscoveryHandler` I don't think I'll move these up after all.

Comment (sanderr) [reply src/inmanta/agent/handler.py:463]: Perhaps `execute` shouldn't exist on the `ABC` then?

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:437]: I don't think this one or `pre` should be marked as `abstractmethod` because that requires concrete classes to override them. We want their default behavior to  be a NOOP, don't we?

Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py:437]: I agree, that's the route I took on my PR implementing the `DiscoveryHandler`, I'll change it here as well

File: src/inmanta/agent/handler.py
Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py:463]: Agreed, and perhaps neither should `deploy`, I'll update this accordingly after feedback on the PR implementing the `DiscoveryHandler`

File: src/inmanta/agent/handler.py
Comment (arnaudsjs) [inline src/inmanta/agent/handler.py:506]:
```suggestion
    A baseclass for classes that handle resources. New handlers are registered with the
```

File: src/inmanta/agent/handler.py
Comment (arnaudsjs) [inline src/inmanta/agent/handler.py:518]: Shouldn't we call super here instead of setting `self._agent`, `self._ioloop` and `self._client` directly?

File: src/inmanta/agent/handler.py
Comment (arnaudsjs) [inline src/inmanta/agent/handler.py:463]: This docstring doesn't really describe the behavior of the method.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:506]: The part about handler registration seems relevant for the top level ABC as well.

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:460]: I think this would be a good time to clean up this type. Either `dict` or `abc.Mapping`. Afaik we don't ever update it so `Mapping` would be best. @arnaudsjs do you know?

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:504]:
```suggestion
class ResourceHandler(HandlerABC[R]):
```
I think. Then update all type annotations of `resources.Resource` to `R` instead. Same for the `CRUDHandler`.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py
Comment (arnaudsjs) [reply src/inmanta/agent/handler.py:460]: I don't think we ever update it.

File: src/inmanta/agent/handler.py
Comment (arnaudsjs) [inline src/inmanta/agent/handler.py:421]:
```suggestion
@stable_api
class HandlerAPI(ABC):
```

File: src/inmanta/agent/handler.py
Comment (arnaudsjs) [inline src/inmanta/agent/handler.py:469]: What is the difference between the `deploy()` and the `execute()` method in the context of a `DiscoveryHandler`?

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py
Comment (wouterdb) [reply src/inmanta/agent/handler.py:469]: dryrun still uses execute instead of deploy

**Review by wouterdb — APPROVED** [review]

File: src/inmanta/agent/handler.py
Comment (arnaudsjs) [reply src/inmanta/agent/handler.py:469]: Ok. What would be the main entrypoint for the `DiscoveryHandler`? Is it `deploy()` or `execute()`. Because I would find `deploy()` a strange name. It doesn't really deploy anything in the end.

**Review by arnaudsjs — APPROVED** [review]

File: src/inmanta/agent/handler.py
Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py:470]: I agree, I'm trying to come up with some clearer phrasing

Comment (sanderr) [reply src/inmanta/agent/handler.py:470]: I added a suggested change to the `deploy` docstring to try and clear this up. What do you think? Anything missing or still unclear there? After we've reached consensus there I think we need to clear up this docstring in the same vein.

**Review by sanderr — COMMENTED** [review]
Releasing comments to enable discussion about "deploy" semantics. Review still in progress.

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:425]:
```suggestion
    Base class describing the interface between the agent and the handler. This class first defines the interface.
    At the end, it also defines a number of utility methods.
```

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:484]: Same question: why is this abstract while still declaring how it interacts with utility methods?

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:501]: I think this is a good time to clarify this docstring. I only know what it means because I've seen it in use.

File: src/inmanta/agent/handler.py
Comment (sanderr) [reply src/inmanta/agent/handler.py:473]: Also it references `ResourceHandler` methods rather than `HandlerAPI` ones. If you update this, please double check that they are all defined on the ABC.

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:571]: I don't really see the relevance of the second sentence. @arnaudsjs @wouterdb do you?

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:547]: I'm not sure if the second part is really relevant for the `get_file` method. It seems like it would be mostly relevant for the upload.

File: src/inmanta/agent/handler.py
Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py:470]: What do you mean exactly with lsm ? (I only found this handler https://github.com/inmanta/lsm/blob/5ba3b8fa9e60b358e311d77ce2d4ab4aa37efe4e/plugins/__init__.py#L1567 which does have an execute method through inheritance)

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:505]: `set_cache` needs a docstring.

File: src/inmanta/agent/handler.py
Comment (sanderr) [reply src/inmanta/agent/handler.py:470]: That's the one I meant, and luckily you're right, I was missing something here.

**Review by sanderr — CHANGES_REQUESTED** [review]
I'm adding a lot of comments to try and clear up docstrings. I'm not intimately familiar with all of them so others may not agree with my suggestions: if you have doubts about any of them I would suggest asking the team before sinking too much time in it. And either way rerequest everyone's review afterwards because these are quite impactful changes.

File: src/inmanta/agent/handler.py
Comment (arnaudsjs) [reply src/inmanta/agent/handler.py:571]: Me neither.

File: src/inmanta/agent/handler.py
Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py]: removed those

File: src/inmanta/agent/handler.py
Comment (Hugo-Inmanta) [inline src/inmanta/agent/handler.py:575]: Had to lift these two as well since they're called from the `deploy` method. Suggestions for better docstrings are welcome

**Comment by Hugo-Inmanta** [comment]
One mypy error popped up, I think it's tied to genericity, I'm not sure if we can ignore it or not:
```diff
+ src/inmanta/agent/handler.py:1008: error: Argument 2 of "execute" is incompatible with supertype "HandlerAPI"; supertype defines the argument type as "Resource"  [override]
```

File: src/inmanta/agent/handler.py
Comment (wouterdb) [reply src/inmanta/agent/handler.py:575]: Why lift the implementation of deploy? 

My understanding is that this class is purely the definition of the interface between agent and handler, so the deploy logic is not part of that?

**Comment by sanderr** [comment]
> One mypy error popped up, I think it's tied to genericity, I'm not sure if we can ignore it or not:

It has to do with the fact that we try to keep some classes non-generic. Ignore it for now but make sure to run mypy again once we decided where to merge this.

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:451]:
```suggestion
        instantiated handler will no longer be used by the agent.
```
Right?

File: src/inmanta/agent/handler.py
Comment (sanderr) [reply src/inmanta/agent/handler.py:575]: Most of what it does is relevant for the generic concept of a resource, isn't it? We could choose to extract some parts of it to the `Resource` subclass, but any resource has dependencies and its handler should generally be executed when those dependencies have successfully deployed.

Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py:575]: As discussed: keep it as is, but unmark HandlerABC stable_api annotation

File: src/inmanta/agent/handler.py
Comment (sanderr) [reply src/inmanta/agent/handler.py:576]: Looks good to me, but @wouterdb is more knowledgeable about it.

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:554]: I just realized that iirc we want to drop this sometime soon (@wouterdb can you confirm?). Perhaps we should declare that it should generally not be overridden?
```suggestion
        Kept for backwards compatibility, new handler implementations should never override this.
```

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py
Comment (Hugo-Inmanta) [reply src/inmanta/agent/handler.py]: > pre and post are concepts of the HandlerAPI

I think they got dropped from the ABC when I merged wouter's suggestion, I put them back in, sorry for the mess.

> Same for the facts() reference and the default implementation.

I lifted those up in the ABC as well

File: src/inmanta/agent/handler.py
Comment (sanderr) [inline src/inmanta/agent/handler.py:583]: I think this should be in the utility methods section. Can you check for the other methods you have lifted in the past few days? I think `pre` and `post` could also be considered utility methods if we define the interface methods as the "agent-handler API interface".

**Review by wouterdb — APPROVED** [review]

**Review by sanderr — COMMENTED** [review]
Looks good to me except for the `Generic` discussion which is blocked until Bart is back.

File: src/inmanta/agent/handler.py
Comment (arnaudsjs) [inline src/inmanta/agent/handler.py:554]:
```suggestion
        :func:`~inmanta.agent.handler.HandlerAPI.facts` to do the actual querying.
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by arnaudsjs — APPROVED** [review]

**Comment by Hugo-Inmanta** [comment]
Still a couple of outstanding mypy issues:
```
+ src/inmanta/agent/handler.py:651: error: Argument 2 to "chain_future" has incompatible type "concurrent.futures._base.Future[T]"; expected "asyncio.futures.Future[Any]"  [arg-type]
+ src/inmanta/agent/handler.py:920: error: Argument 2 to "read_resource" of "CRUDHandler" has incompatible type "PurgeableResource"; expected "TPurgeableResource"  [arg-type]
+ src/inmanta/agent/handler.py:921: error: Argument 2 to "calculate_diff" of "CRUDHandler" has incompatible type "PurgeableResource"; expected "TPurgeableResource"  [arg-type]

```
Ultimately tied to the typing of Resource https://github.com/inmanta/inmanta-core/blob/cffef5db5d3e21fc54c10a30c1cee56d775eb2db/src/inmanta/resources.py#L425 let me know if you see a solution for those

**Review by sanderr — CHANGES_REQUESTED** [review]
Looks mostly good. A few minor notes/suggestions and a few notes for the follow-up ticket.

> let me know if you see a solution for those

- For the first, I don't think this is actually new, but it does look like an actual type error to me. I think using `asyncio.Future` instead of `concurrent.futures.Future` should resolve that. That is unless @wouterdb knows of any reason we would want to violate this?

- For the second and third: drop the assertions, and update `Resource.clone` and `Resource.deserialize` to return the type of the class rather than `Resource`.

**Comment by arnaudsjs** [comment]
Regarding the typing issue on `chain_future`: This seems to be a typing issue in Tornado. The `chain_future` is annotated to accept a `asyncio.Future` instance, but according to the implementation and the documentation of the method it accepts all types of Futures (asyncio.Future and concurrent.futures.Future). I created [a ticket](https://github.com/tornadoweb/tornado/issues/3314) for this.

**Comment by arnaudsjs** [comment]
@sanderr I still have the following type error left:

```
src/inmanta/resources.py:435: error: Incompatible return value type (got "Resource", expected "T")  [return-value]
```

I don't really see how that can be fixed. `Resource.deserialize()` is a class method. Its return type depends on what `resource.get_class()` returns and that can be any subclass of Resource.

**Pull request closed without merging**
